### PR TITLE
Smart contract review

### PR DESCRIPTION
**Pull request by bsasbon-np**

### CODE REVIEW

File: contract/lib.rs [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+        #[ink(message)]
+        pub fn award(&mut self, recipient: AccountId, amount: u64) -> Result<(), ContractError> {
+            let sender = self.env().caller();
+
+            if sender == recipient {
+                return Err(ContractError::CannotAwardYourself);
+            }
+
+            self.check_sender_and_recipient(&sender, &recipient)?;
```
Comment (bsasbon-np) [inline contract/lib.rs]: Rename to "validate_is_member" or "check_valid_member"

File: contract/lib.rs [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+
+            let sender_points = self.points_by_sender.get(sender).unwrap_or(0);
```
Comment (bsasbon-np) [inline contract/lib.rs]: Rename variable to "sender_used_points"

File: contract/lib.rs [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+            }
+        }
+
+        #[ink(message)]
+        pub fn set_admin(&mut self, new_admin: AccountId) -> Result<(), ContractError> {
+            self.check_admin()?;
+
+            if !self.is_member(&new_admin) {
+                self.add_member(new_admin)?;
+            }
```
Comment (bsasbon-np) [inline contract/lib.rs:89]: Admin might not necessarily be a member, I think we can remove this part

File: contract/lib.rs [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+            self.check_admin()?;
+
+            if self.max_points_per_sender == 0 {
+                return Err(ContractError::MaxPointsPerSenderCannotBeZero);
+            }
+
+            self.max_points_per_sender = max_points_per_sender;
+
+            Ok(())
+        }
```
Comment (bsasbon-np) [inline contract/lib.rs:143]: Should we add a max value of points per sender? Something like 500

File: contract/lib.rs [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+
+        #[ink(message)]
+        pub fn award(&mut self, recipient: AccountId, amount: u64) -> Result<(), ContractError> {
```
Comment (bsasbon-np) [inline contract/lib.rs]: I would rename the "amount" param to "points"

File: contract/lib.rs [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+            self.validate_award_inputs(sender, recipient, amount)?;
+
+            let sender_used_points = self.points_by_sender.get(sender).unwrap_or(0);
+            self.validate_sender_points(sender_used_points, amount)?;
+            self.points_by_sender
+                .insert(sender, &(sender_used_points.checked_add(amount).unwrap()));
+
+            let recipient_points = self.points_by_recipient.get(recipient).unwrap_or(0);
+            self.points_by_recipient
+                .insert(recipient, &(recipient_points.checked_add(amount).unwrap()));
```
Comment (bsasbon-np) [inline contract/lib.rs]: To make it more explicit I would take out lines like:

`&(sender_used_points.checked_add(amount).unwrap())`
`&(recipient_points.checked_add(amount).unwrap())` and


To their own line where we name variables like "sender_used_points_result" and "recipients_points_result"

File: contract/lib.rs [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+                amount,
+            });
+
+            self.update_senders_and_recipients(sender, recipient);
+
+            Ok(())
+        }
+
+        #[ink(message)]
+        pub fn distribute_rewards(&mut self, amount: Option<Balance>) -> Result<(), ContractError> {
```
Comment (bsasbon-np) [inline contract/lib.rs]:
```suggestion
        pub fn distribute_rewards(&mut self, amount_to_distribute: Option<Balance>) -> Result<(), ContractError> {
```

File: contract/lib.rs [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+                let recipient_points = self.points_by_recipient.get(recipient).unwrap();
+                self.points_by_recipient.remove(recipient);
+
+                let recipient_reward: Balance = (recipient_points as u128)
+                    .checked_mul(total_rewards)
+                    .unwrap()
+                    .checked_div(self.total_points as u128)
+                    .unwrap();
+
+                // ToDo: Test the expect
```
Comment (bsasbon-np) [inline contract/lib.rs]: Remove comment

File: contract/lib.rs [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+                    points: recipient_points,
+                });
+            }
+
+            self.reset_points();
+
+            Ok(())
+        }
+
+        pub fn validate_rewards_recipients(&self) -> Result<(), ContractError> {
```
Comment (bsasbon-np) [inline contract/lib.rs]:
```suggestion
        pub fn validate_recipients_exist(&self) -> Result<(), ContractError> {
```

File: contract/lib.rs [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+            Ok(())
+        }
+
+        #[ink(message)]
+        pub fn distribute_rewards(&mut self, amount: Option<Balance>) -> Result<(), ContractError> {
+            self.check_admin()?;
+
+            let total_rewards = self.validate_reward(amount)?;
+
+            self.validate_rewards_recipients()?;
```
Comment (bsasbon-np) [inline contract/lib.rs]: I would validate this before validating the amount

File: contract/tests.rs [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+    // Assert
+    assert_eq!(result_award_bob_ok, Ok(()));
+    assert_eq!(result_award_frank_ok, Ok(()));
+    assert_eq!(contract.total_points, 5);
+    assert_eq!(contract.recipients.len(), 2);
+    assert_eq!(contract.senders.len(), 1);
+}
+
+#[ink::test]
+fn award_with_null_points_amount_fails_() {
```
Comment (bsasbon-np) [inline contract/tests.rs]:
```suggestion
fn award_with_null_points_amount_fails() {
```

File: contract/tests.rs [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+
+    // Act
+    let result_fails = contract.award(accounts.alice, 1);
+
+    // Assert
+    assert_eq!(result_fails, Err(ContractError::CannotAwardYourself));
+}
+
+#[ink::test]
+fn award_with_points_amount_exceeding_the_maximum() {
```
Comment (bsasbon-np) [inline contract/tests.rs]:
```suggestion
fn award_with_points_amount_exceeding_the_maximum_fails() {
```

File: contract/tests.rs [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+
+    // Act
+    let result_fails = contract.award(accounts.bob, 10);
+
+    // Assert
+    assert_eq!(result_fails, Err(ContractError::MaxPointsPerSenderExceeded));
+}
+
+#[ink::test]
+fn distribute_rewards_works() {
```
Comment (bsasbon-np) [inline contract/tests.rs:151]: Here we are missing the asserts where we check that each member got the expected amount of funds

File: contract/lib.rs (hunk 3 above)
Comment (pcorrado-np) [reply contract/lib.rs:89]: Are you saying that an admin can be external to the DAO? Sounds weird..

File: contract/lib.rs (hunk 4 above)
Comment (pcorrado-np) [reply contract/lib.rs:143]: The constructor has 10 as a default value.

File: contract/tests.rs (hunk 13 above)
Comment (pcorrado-np) [reply contract/tests.rs:151]: We have to do an E2E test to check that

Comment (pcorrado-np) [reply contract/tests.rs:151]: :arrow_forward: Pending to do E2E tests

**Pull request merged**